components/core/src/clp/streaming_compression/lzma/Compressor.cpp

```diff
@@ -0,0 +1,248 @@
+
+    // Detach output buffer from LZMA stream
+    m_compression_stream.next_out = nullptr;
+    m_compression_stream.avail_out = 0;
+
+    m_compressed_stream_file_writer = nullptr;
+}
+
+auto Compressor::write(char const* data, size_t data_length) -> void {
+    if (nullptr == m_compressed_stream_file_writer) {
+        throw OperationFailed(ErrorCode_NotInit, __FILENAME__, __LINE__);
+    }
+
+    if (0 == data_length) {
+        // Nothing needs to be done because we do not need to compress anything
+        return;
+    }
+
+    if (nullptr == data) {
+        throw OperationFailed(ErrorCode_BadParam, __FILENAME__, __LINE__);
+    }
+
+    // Attach input data to LZMA stream
+    m_compression_stream.next_in = clp::size_checked_pointer_cast<uint8_t const>(data);
+    m_compression_stream.avail_in = data_length;
+
+    while (m_compression_stream.avail_in > 0) {
+        encode_lzma_once();
+    }
+
+    // All input data have been encoded so detach input data
+    m_compression_stream.next_in = nullptr;
+
+    m_uncompressed_stream_pos += data_length;
+}
+
+auto Compressor::flush() -> void {
+    flush_lzma(LZMA_SYNC_FLUSH);
+}
+
+auto Compressor::try_get_pos(size_t& pos) const -> ErrorCode {
+    if (nullptr == m_compressed_stream_file_writer) {
+        return ErrorCode_NotInit;
+    }
+
+    pos = m_uncompressed_stream_pos;
+    return ErrorCode_Success;
+}
+
+auto Compressor::encode_lzma_once() -> void {
+    if (0 == m_compression_stream.avail_in) {
+        return;
+    }
+
+    if (0 == m_compression_stream.avail_out) {
+        flush_stream_output_block_buffer();
+    }
+
+    auto const rc = lzma_code(&m_compression_stream, LZMA_RUN);
+    switch (rc) {
+        case LZMA_OK:
+            break;
+        case LZMA_BUF_ERROR:  // No encoding progress can be made
+            SPDLOG_ERROR("LZMA compressor input stream is corrupt.");
+            throw OperationFailed(ErrorCode_Failure, __FILENAME__, __LINE__);
+        default:
+            SPDLOG_ERROR("lzma_code() returned an unexpected value - {}.", static_cast<int>(rc));
+            throw OperationFailed(ErrorCode_Failure, __FILENAME__, __LINE__);
+    }
+}
+
+auto Compressor::flush_lzma(lzma_action flush_action) -> void {
+    if (false == is_flush_action(flush_action)) {
+        SPDLOG_ERROR(
+                "lzma_code() supplied with invalid flush action - {}.",
+                static_cast<int>(flush_action)
+        );
+        throw OperationFailed(ErrorCode_BadParam, __FILENAME__, __LINE__);
+    }
+
+    bool flushed{false};
+    while (false == flushed) {
+        auto const rc = lzma_code(&m_compression_stream, flush_action);
+        switch (rc) {
+            case LZMA_OK:
+                break;
+            case LZMA_STREAM_END:
+                // NOTE: this might not be true when multithreaded encoder is
+                // used with LZMA_FULL_BARRIER. For now, we skip this check.
+                flushed = true;
+                break;
```

Comment:
2 questions:
1. is LZMA_OK a possible return code? The handling looks weird as it breaks from the loop leaving flushed still set to false. 
2. flushed is not used anywhere. in LZMA_STREAM_END case, since you break anyway, setting flushed = true doesn't have a real impact.